modules: hal_nordic: nrfx: update API version to 3.2.0

nika-nordic opened the pull request:
Updated API version enables multi-instance GPIOTE driver.

Additionally obsolete symbol that was used to specify API version in the past was removed.

nika-nordic commented:
@anangl CI seems to be green, can we proceed with review and merging?

nika-nordic commented:
Rebased the PR. Hopefully now the unrelated:
> FAILED: zephyr/subsys/net/CMakeFiles/subsys__net.dir/lib/tls_credentials/tls_credentials_trusted.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.16.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DNRF5340_XXAA_APPLICATION -DNRF_SKIP_FICR_NS_COPY_TO_RAM -DNRF_TRUSTZONE_NONSECURE -DPICOLIBC_LONG_LONG_PRINTF_SCANF -DTC_RUNID=9473c3a39eb23922342372ed882ae5f4 -D_FORTIFY_SOURCE=1 -D_POSIX_C_SOURCE=200809 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/subsys/net/lib/tls_credentials/. -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/thingy53_nrf5340_cpuapp_ns/tests/net/lib/tls_credentials/net.tls_credentials.trusted_tfm/zephyr/include/generated -I/__w/zephyr/zephyr/soc/arm/nordic_nrf/nrf53 -I/__w/zephyr/zephyr/soc/arm/nordic_nrf/common/. -I/__w/zephyr/zephyr/subsys/testsuite/include -I/__w/zephyr/zephyr/subsys/testsuite/ztest/include -I/__w/zephyr/zephyr/subsys/usb/device -I/__w/zephyr/zephyr/subsys/net/lib/dns/. -I/__w/zephyr/zephyr/drivers/usb/common/nrf_usbd_common/. -I/__w/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/__w/zephyr/zephyr/modules/cmsis/. -I/__w/zephyr/modules/hal/nordic/nrfx -I/__w/zephyr/modules/hal/nordic/nrfx/drivers/include -I/__w/zephyr/modules/hal/nordic/nrfx/mdk -I/__w/zephyr/zephyr/modules/hal_nordic/nrfx/. -I/__w/zephyr/zephyr/twister-out/thingy53_nrf5340_cpuapp_ns/tests/net/lib/tls_credentials/net.tls_credentials.trusted_tfm/tfm/generated/interface/include -isystem /__w/zephyr/zephyr/lib/libc/common/include -Wshadow -fno-strict-aliasing -Werror -Os -imacros /__w/zephyr/zephyr/twister-out/thingy53_nrf5340_cpuapp_ns/tests/net/lib/tls_credentials/net.tls_credentials.trusted_tfm/zephyr/include/generated/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m33 -mthumb -mabi=aapcs -mfp16-format=ieee -mtp=soft --sysroot=/opt/toolchains/zephyr-sdk-0.16.3/arm-zephyr-eabi/arm-zephyr-eabi -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/tests/net/lib/tls_credentials=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections --specs=picolibc.specs -std=c99 -MD -MT zephyr/subsys/net/CMakeFiles/subsys__net.dir/lib/tls_credentials/tls_credentials_trusted.c.obj -MF zephyr/subsys/net/CMakeFiles/subsys__net.dir/lib/tls_credentials/tls_credentials_trusted.c.obj.d -o zephyr/subsys/net/CMakeFiles/subsys__net.dir/lib/tls_credentials/tls_credentials_trusted.c.obj -c /__w/zephyr/zephyr/subsys/net/lib/tls_credentials/tls_credentials_trusted.c
/__w/zephyr/zephyr/subsys/net/lib/tls_credentials/tls_credentials_trusted.c:12:10: fatal error: psa/protected_storage.h: No such file or directory
   12 | #include <psa/protected_storage.h>

will be fixed

anangl commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -85,6 +85,14 @@ DT_FOREACH_PROP_ELEM(MATRIX_NODE, pixel_mapping, CHECK_PIXEL)
+#define GPIOTE_PHANDLE(pin) DT_PHANDLE(GPIO_NODE(pin), gpiote_instance)
+#define GPIOTE_IDX(pin)     DT_PROP(GPIOTE_PHANDLE(pin), instance)
+
+static nrfx_gpiote_t gpiote = NRFX_GPIOTE_INSTANCE(GPIOTE_IDX(col_gpios));
```
This should not assume a single GPIOTE instance. I think that a similar scheme as in `pwm_nrf_sw.c` should be used instead.

masz-nordic replied:
Done

masz-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -7,6 +7,7 @@
 #define ADC_CONTEXT_USES_KERNEL_TIMER
 #include "adc_context.h"
 #include <hal/nrf_saadc.h>
+#include <nrfx_saadc.h>
```
`NRFX_SAADC_SAMPLES_TO_BYTES` is used.

anangl replied:
Yes, and in my opinion this is wrong.

masz-nordic replied:
Dropped whole SAADC

masz-nordic commented on modules/hal_nordic/nrfx/nrfx_config_common.h:
```diff
@@ -18,7 +18,7 @@
 /** @brief Symbol specifying minor version of the nrfx API to be used. */
 #ifndef NRFX_CONFIG_API_VER_MINOR
-#define NRFX_CONFIG_API_VER_MINOR 0
+#define NRFX_CONFIG_API_VER_MINOR 2
```
That would make that one commit huge, combining multiple areas - is it really better?

tmon-nordic replied:
Yes, because otherwise the code does not build. Any commit that does not build is significantly slowing down bisect because you have to skip one, or more, commits which significantly increases the required number of checks. The required number of checks, when all commits in history do build correctly, is log2(N) where N is the number of commits in between good and bad revisions.

tmon-nordic replied:
The commits should really be:

- Update nrfx API version with all drivers adjusted for new API. No actual functional changes (if possible).
- Any subsequent commits are the functional changes related to things that were not possible with earlier nrfx version.

masz-nordic commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -131,11 +131,12 @@
 			reg = <0x40006000 0x1000>;
 			interrupts = <6 NRF_DEFAULT_IRQ_PRIORITY>;
 			status = "disabled";
+			instance = <0>;
```
Done

masz-nordic commented on soc/arm/nordic_nrf/validate_base_addresses.c:
```diff
@@ -140,7 +140,16 @@ CHECK_DT_REG(ficr, NRF_FICR);
 CHECK_DT_REG(gpio0, NRF_P0);
 CHECK_DT_REG(gpio1, NRF_P1);
-CHECK_DT_REG(gpiote, NRF_GPIOTE);
+#if defined(NRF_GPIOTE)
+CHECK_DT_REG(gpiote0, NRF_GPIOTE);
+#else
+CHECK_DT_REG(gpiote0, NRF_GPIOTE0);
+#endif
```
This still causes compilation error on `nrf5340_cpuapp` as it has both `gpiote` and `gpiote0` nodes defined, but only `NRF_GPIOTE0`. Adding `#define NRF_GPIOTE NRF_GPIOTE0` would fix it but I'm not sure if this is a better solution than what we have now.

masz-nordic commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -63,8 +62,13 @@ struct spi_nrfx_config {
 };
 
+#ifdef CONFIG_SOC_NRF52832_ALLOW_SPIM_DESPITE_PAN_58
+static const nrfx_gpiote_t gpiote = NRFX_GPIOTE_INSTANCE(0);
```
Done

anangl commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -340,7 +354,7 @@ static void prepare_pixel_pulse(const struct device *dev,
 	/* First timer channel is used for timing the period of pulses. */
 	nrf_timer_cc_set(dev_config->timer, 1 + channel_idx, pulse);
-	NRF_GPIOTE->CONFIG[dev_data->gpiote_ch[channel_idx]] = gpiote_cfg;
+	gpiote[channel_idx].p_reg->CONFIG[dev_data->gpiote_ch[channel_idx]] = gpiote_cfg;
```
This cannot be done this way for the general case. GPIOTE channels here may be used with various column pins, not necessarily with the same ones that were used to fill the `gpiote` array at particular indexes. See https://github.com/zephyrproject-rtos/zephyr/blob/5a2b29169780ec9cc69865f23bed6e1ca76bea07/drivers/display/display_nrf_led_matrix.c#L334-L335
I think we should rather limit the driver to use one GPIOTE instance and ensure at build time that all column GPIOs can use that instance. I'll dig into this further and propose something specific.

jaz1-nordic committed:
boards: arm: nrf5340_audio_dk_cpunet_reset: switch to HAL

Change the GPIOTE driver to HAL to prevent instantiation
issues with a multi-instance GPIOTE driver.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic committed:
dts: arm: add new gpiote instances definition

Added GPIOTE0, GPIOTE1 instances for legacy devices,
GPIOTE20, GPIOTE30 for Moonlight and GPIOTE130,
GPIOTE131 instances for Haltium.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic committed:
boards: arm: set gpiote status to okay as default

After adding new GPIOTE instances, there is a need
to enable the instance for individual boards.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

anangl commented on samples/boards/nrf/nrfx/Kconfig:
```diff
@@ -2,9 +2,19 @@
 
 config NRFX_DPPI
-	default HAS_HW_NRF_DPPIC
+	default $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF_DPPIC))
 
 config NRFX_PPI
-	default HAS_HW_NRF_PPI
+	default $(dt_has_compat,$(DT_COMPAT_NORDIC_NRF_PPI))
```
This should not be changed in this commit and PR. It's unrelated.

anangl commented on scripts/kconfig/kconfigfunctions.py:
Changes in this file should be dropped.

anangl commented on soc/arm/nordic_nrf/common/soc_nrf_common.h:
```diff
@@ -149,6 +149,14 @@
 		    (NRF_DT_GPIOS_TO_PSEL(node_id, prop)),	\
 		    (default_value))
 
+#define NRF_DT_GPIOTE_INST_BY_IDX(node_id, prop, idx)			\
+	DT_PROP(DT_PHANDLE(DT_GPIO_CTLR_BY_IDX(node_id, prop, idx),	\
+			   gpiote_instance),				\
+		instance)
+
+#define NRF_DT_GPIOTE_INST(node_id, prop)				\
+	NRF_DT_GPIOTE_INST_BY_IDX(node_id, prop, 0)
```
These macros need descriptions.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -55,6 +56,10 @@
 	 GPIOTE_CONFIG_OUTINIT_High :			\
 	 GPIOTE_CONFIG_OUTINIT_Low)
 
+#define GPIO_NODE(node, pin, idx)      DT_GPIO_CTLR_BY_IDX(node, pin, idx)
+#define GPIOTE_PHANDLE(node, pin, idx) DT_PHANDLE(GPIO_NODE(node, pin, idx), gpiote_instance)
+#define GPIOTE_IDX(node, pin, idx)     DT_PROP(GPIOTE_PHANDLE(node, pin, idx), instance)
```
`NRF_DT_GPIOTE_INST_BY_IDX` should be used instead.

anangl commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_dppi.h:
```diff
@@ -180,6 +180,13 @@ static inline void hal_trigger_rateoverride_ppi_config(void)
 
+#if defined(HAL_RADIO_GPIO_HAVE_PA_PIN) || defined(HAL_RADIO_GPIO_HAVE_LNA_PIN) ||                 \
+	defined(HAL_RADIO_FEM_IS_NRF21540)
+#define GPIO_NODE(node, pin, idx)      DT_GPIO_CTLR_BY_IDX(node, pin, idx)
+#define GPIOTE_PHANDLE(node, pin, idx) DT_PHANDLE(GPIO_NODE(node, pin, idx), gpiote_instance)
+#define GPIOTE_IDX(node, pin, idx)     DT_PROP(GPIOTE_PHANDLE(node, pin, idx), instance)
```
ditto

anangl commented on samples/boards/nrf/nrfx/src/main.c:
```diff
@@ -18,8 +18,17 @@
 
-#define INPUT_PIN	DT_GPIO_PIN(DT_ALIAS(sw0), gpios)
-#define OUTPUT_PIN	DT_GPIO_PIN(DT_ALIAS(led0), gpios)
+#define INPUT_PIN   NRF_DT_GPIOS_TO_PSEL(DT_ALIAS(sw0), gpios)
+#define OUTPUT_PIN  NRF_DT_GPIOS_TO_PSEL(DT_ALIAS(led0), gpios)
```
This change is also unrelated and should not be done in this PR.

anangl commented on samples/boards/nrf/nrfx/src/main.c:
```diff
@@ -18,8 +18,17 @@
+#define INPUT_PIN   NRF_DT_GPIOS_TO_PSEL(DT_ALIAS(sw0), gpios)
+#define OUTPUT_PIN  NRF_DT_GPIOS_TO_PSEL(DT_ALIAS(led0), gpios)
+
+#define GPIOTE_INST NRF_DT_GPIOTE_INST(DT_ALIAS(sw0), gpios)
+#define GPIOTE_NODE DT_NODELABEL(_CONCAT(gpiote, GPIOTE_INST))
```
```suggestion
#define GPIOTE_INST	NRF_DT_GPIOTE_INST(DT_ALIAS(sw0), gpios)
#define GPIOTE_NODE	DT_NODELABEL(_CONCAT(gpiote, GPIOTE_INST))
```

anangl commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -20,13 +21,13 @@
-#define PWM_NODE     DT_PHANDLE(MATRIX_NODE, pwm)
-#define ROW_COUNT    DT_PROP_LEN(MATRIX_NODE, row_gpios)
-#define COL_COUNT    DT_PROP_LEN(MATRIX_NODE, col_gpios)
-#define GROUP_SIZE   DT_PROP(MATRIX_NODE, pixel_group_size)
+#define MATRIX_NODE DT_INST(0, nordic_nrf_led_matrix)
+#define TIMER_NODE  DT_PHANDLE(MATRIX_NODE, timer)
+#define USE_PWM     DT_NODE_HAS_PROP(MATRIX_NODE, pwm)
+#define PWM_NODE    DT_PHANDLE(MATRIX_NODE, pwm)
+#define ROW_COUNT   DT_PROP_LEN(MATRIX_NODE, row_gpios)
+#define COL_COUNT   DT_PROP_LEN(MATRIX_NODE, col_gpios)
+#define GROUP_SIZE  DT_PROP(MATRIX_NODE, pixel_group_size)
```
Please drop this change.

anangl commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -408,20 +431,36 @@ static const struct gpio_driver_api gpio_nrfx_drv_api_funcs = {
 };
 
+#define GPIOTE_PHANDLE(id) DT_INST_PHANDLE(id, gpiote_instance)
+#define GPIOTE_IDX(id)     DT_PROP(GPIOTE_PHANDLE(id), instance)
+
+#define GPIOTE_INSTANCE(id)                                 \
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(id, gpiote_instance), \
+		    (NRFX_GPIOTE_INSTANCE(GPIOTE_IDX(id))),         \
```
```suggestion
#define GPIOTE_INST(id)    DT_PROP(GPIOTE_PHANDLE(id), instance)

#define GPIOTE_INSTANCE(id)					\
	COND_CODE_1(DT_INST_NODE_HAS_PROP(id, gpiote_instance),	\
		    (NRFX_GPIOTE_INSTANCE(GPIOTE_INST(id))),	\
```

anangl commented on soc/arm/nordic_nrf/common/soc_nrf_common.h:
```diff
@@ -149,6 +149,52 @@
+ *             instance = <20>;
+ *     };
+ *
+ *     gpio0 {
+ *             gpiote-instance = <&gpiote0>;
+ *     }
+ *
+ *     gpio1 {
+ *             gpiote-instance = <&gpiote20>;
+ *     }
```
```suggestion
 *     &gpio0 {
 *             gpiote-instance = <&gpiote0>;
 *     };
 *
 *     &gpio1 {
 *             gpiote-instance = <&gpiote20>;
 *     };
```

anangl commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -444,6 +446,7 @@ static int spi_nrfx_init(const struct device *dev)
 		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(SPI(idx)),		       \
 		.wake_pin = NRF_DT_GPIOS_TO_PSEL_OR(SPI(idx), wake_gpios,      \
 						    WAKE_PIN_NOT_USED),	       \
+		.wake_gpiote = WAKE_GPIOTE_INSTANCE(SPI(idx)), \
```
```suggestion
		.wake_gpiote = WAKE_GPIOTE_INSTANCE(SPI(idx)),		       \
```

anangl commented on drivers/display/display_nrf_led_matrix.c:
```diff
@@ -530,6 +541,8 @@ static const struct display_drv_config instance_config = {
 #if USE_PWM
 	.pwm = (NRF_PWM_Type *)DT_REG_ADDR(PWM_NODE),
+#else
+	.gpiote = NRF_DT_GPIOTE_INST_BY_IDX(MATRIX_NODE, col_gpios, 0),
```
```suggestion
	.gpiote = NRFX_GPIOTE_INSTANCE(
			NRF_DT_GPIOTE_INST_BY_IDX(MATRIX_NODE, col_gpios, 0)),
```

anangl commented on dts/bindings/gpio/nordic,nrf-gpio.yaml:
This change should be in commit 11a93ece9058a6e3f3e5e2d3ebae51d475bd0aa7.

anangl commented on dts/bindings/gpio/nordic,nrf-gpiote.yaml:
This one as well.

anangl commented on soc/arm/nordic_nrf/common/soc_nrf_common.h:
Also this one fits better there.

jaz1-nordic committed:
modules: hal_nordic: nrfx: update API version to 3.2.0

Updated API version enables multi-instance GPIOTE driver.
Additionally obsolete symbol that was used to specify
API version in the past was removed.
Affected drivers have been adjusted and appropriate changes
in affected files have been made.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>
